### PR TITLE
feat:  CCIPConfig, adding exhaustive checks for p2pIds and bootstrapP2PIds

### PR DESCRIPTION
**Pull request by 0xsuryansh**
## Motivation
In `CCIPConfig._validateConfig` we try to do an exhaustive amount of validations, 
the following validations remained
1.  Check whether a duplicate bootstrap P2P id, P2P id, signer, or transmitter was passed in.
2. Check that p2p ids in cfg.bootstrapP2PIds are included in cfg.p2pIds

Ticket [CCIP-2595](https://smartcontract-it.atlassian.net/browse/CCIP-2595)
## Solution
Created a library `CCIPConfigArraysValidation` to support these sort, duplicate and subset check in `CCIPConfig`.
This library can also be used in the optimisation of `CCIPConfigArraysValidation._ensureInRegistry`

Regarding 1 : Since each p2pId at index i corresponds to the signer at index i both arrays can't be sorted at the same time.
Hence, we are going the enforce that p2pIds and bootstrapP2PIds are sent sorted on function calls to in `CCIPConfig ` contract. This means that these exhaustive checks for ` bytes[] signers` and ` bytes[] transmitters` needs to be done in off-chain tooling.

[CCIP-2595]: https://smartcontract-it.atlassian.net/browse/CCIP-2595?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol [hunk 1]
```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.24;
+
+library CCIPConfigArraysValidation {
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol]: suggestion: this looks ccip-agnostic, and generic enough to be placed in [contracts/shared](https://github.com/smartcontractkit/ccip/tree/ccip-develop/contracts/src/v0.8/shared) (we typically first merge this in [chainlink core repo](https://github.com/smartcontractkit/chainlink) and cherry-pick it back)

We could just name it `ArrayValidationUtils` or `SortedArrayValidationUtils` and expose it for broader re-use

@RensR thoughts?

File: contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol [hunk 2]
```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.24;
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol]: suggestion: I think we can just use `^0.8.0` and not restrict it to a specific version since this is a lib

File: contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+library CCIPConfigArraysValidation {
+    // errors
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol]: nit: no need for the header comment

File: contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    error ArrayEmpty();
+    error NotSorted(bytes32[] array);
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol]: suggestion: since this is a library that could be used in many different flows, we should name them more specifically:

* `ArrayNotSorted`, `ArrayHasDuplicates`, `ArrayNotASubsetOf`

File: contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    error HasDuplicates(bytes32[] array);
+    error NotSubset(bytes32[] smaller, bytes32[] larger);
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol]: suggestion: I recon we can just have the `NotSorted` and `HasDuplicates` errors, that should be enough context for all inputs.

I'm also wondering - do we need additional specific errors, as opposed to having a single `SortedArrayValidationFailure`? (since it adds to the contract size)

File: contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    function _checkSortedNoDuplicatesAndSubset(bytes32[] memory a, bytes32[] memory b) internal pure {
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol]: suggestion: we should add natspec formatted docs, see: https://github.com/smartcontractkit/ccip/blob/ccip-develop/contracts/src/v0.8/ccip/libraries/RateLimiter.sol#L39-L45

File: contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        }
+
+        // If not all elements in 'a' were matched
+        if (i < a.length) {
+            revert NotSubset(a, b);
+        }
+    }
+
+    // Helper function to check if array is sorted and has no duplicates
+    function _checkSortedAndDuplicates(bytes32[] memory array) private pure {
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol]: nit: we could rename it more explicitly, e.g. `_checkArraySortedWithUniqueElements`

File: contracts/src/v0.8/ccip/capability/CCIPConfig.sol [hunk 8]
```diff
@@ -91,6 +92,7 @@ contract CCIPConfig is ITypeAndVersion, ICapabilityConfiguration, OwnerIsCreator
     uint8 F; //                       | The "big F" parameter for the role DON.
     uint64 offchainConfigVersion; // ─╯ The version of the offchain configuration.
     bytes offrampAddress; // The remote chain offramp address.
+    //NOTE: bootstrapP2PIds and p2pIds should be sent as sorted arrays
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/capability/CCIPConfig.sol]:
```suggestion
    // NOTE: bootstrapP2PIds and p2pIds should be sent as sorted arrays
```

File: contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol (hunk 6 above)
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol]: suggestion: we could simplify the name to `_checkIsValidUniqueSubset(bytes32[] subset, bytes32[] superset)`

We should rename `a` / `b` to more descriptive var names as well, I suppose `subset` and `superset` would be clear

File: contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        for (uint256 i = 1; i < array.length; ++i) {
+            if (array[i] < array[i - 1]) {
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol]: suggestion: inline with prev comment, we could collapse `NotSorted` and `HasDuplicates` to one error, e.g. `NotASortedSet`. This would also be more contract & gas size efficient:

```solidity
if (array[i] <= array[i - 1]) {
  revert NotASortedSet(array);
}
```

File: contracts/src/v0.8/ccip/test/capability/CCIPConfig.t.sol [hunk 10]
```diff
@@ -42,11 +43,30 @@ contract CCIPConfigSetup is Test {
     return subset;
   }
 
+  function _sort(bytes32[] memory arr, int256 left, int256 right) private pure {
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/test/capability/CCIPConfig.t.sol:48]: suggestion: we can use [OpenZeppelin's sort lib](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/Arrays.sol#L29) instead

File: contracts/src/v0.8/ccip/test/capability/CCIPConfig.t.sol [hunk 11]
```diff
@@ -6,6 +6,7 @@ import {Test} from "forge-std/Test.sol";
 import {CCIPConfig} from "../../capability/CCIPConfig.sol";
 import {ICapabilitiesRegistry} from "../../capability/interfaces/ICapabilitiesRegistry.sol";
 import {CCIPConfigHelper} from "../helpers/CCIPConfigHelper.sol";
+import {CCIPConfigArraysValidation} from "../../capability/CCIPConfigArraysValidation.sol";
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/test/capability/CCIPConfig.t.sol:11]: suggestion: we should add an independent test file for `CCIPConfigArraysValidation` since it can be treated as a separate library.

I would also suggest writing a [Foundry fuzz test](https://book.getfoundry.sh/forge/fuzz-testing) which would test different randomized inputs, to make sure to catch any unknown edge cases

File: contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol (hunk 1 above)
Comment (0xsuryansh) [reply contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol]: Yeah, I had posted the same question on pod onchain. I wanted to move it to `contract/shared` and make it generic

File: contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+pragma solidity 0.8.24;
+
+library CCIPConfigArraysValidation {
+  // errors
+  error ArrayEmpty();
+  error NotSorted(bytes32[] array);
+  error HasDuplicates(bytes32[] array);
+  error NotSubset(bytes32[] smaller, bytes32[] larger);
+
+  function _checkSortedNoDuplicatesAndSubset(bytes32[] memory a, bytes32[] memory b) internal pure {
```
Comment (makramkd) [inline contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol]: This needs a docstring, what is checked for being a subset of what?

File: contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol (hunk 5 above)
Comment (0xsuryansh) [reply contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol]: Yes, having specific error add to the contract size, but this can help better in debugging, also, since I am going to make this a share util, it'll be better to provide users with custom errors, 

Although, one optimisation I see is that we can create an enum

```
enum SortedArrayError {
    Empty,
    NotSorted,
    HasDuplicates,
    NotASubset
}
```
and use this


`error SortedArrayValidationFailure(SortedArrayError)`

Let me know your thoughts

File: contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol (hunk 12 above)
Comment (0xsuryansh) [reply contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol]: Added https://github.com/smartcontractkit/ccip/pull/1129/commits/1065995f40c00dca4e8b666c2981d2052ca6ea16

File: contracts/src/v0.8/ccip/test/capability/CCIPConfig.t.sol (hunk 11 above)
Comment (0xsuryansh) [reply contracts/src/v0.8/ccip/test/capability/CCIPConfig.t.sol:11]: Made changes here https://github.com/smartcontractkit/ccip/commit/1065995f40c00dca4e8b666c2981d2052ca6ea16

Created a new library `SortedSetValidationUtil` with new errors 
```
error EmptySet();
error NotASortedSet(bytes32[] set);
 error NotASubset(bytes32[] subset, bytes32[] superset);
```

File: contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol (hunk 2 above)
Comment (0xsuryansh) [reply contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol]: Done

File: contracts/src/v0.8/ccip/test/capability/CCIPConfig.t.sol (hunk 10 above)
Comment (0xsuryansh) [reply contracts/src/v0.8/ccip/test/capability/CCIPConfig.t.sol:48]: The OZ version we are using does not seem to have this

File: contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol (hunk 5 above)
Comment (elatoskinas) [reply contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol]: In my opinion having 4 different errors for one input is not that useful, if we know exactly which array resulted in an error, we can easily deduct the issue by checking the function sig, which should say something like:

```solidity
/// @param bootstrapIds List of bootstrap IDs to set.
/// The bootstrap IDs should be a non-empty subset of all p2p IDs, contain no duplicates and be sorted
```

Even if there was a specific error, the person that sets the config will be very likely to re-read the doc to make sure they haven't missed any other assumptions to prevent another misconfig.

So in my opinion two errors should be enough:

```
error NotANonEmptySortedSet(bytes32[] inputSubset);
error InvalidSortedSubset(bytes32[] inputSubset, bytes32[] superset);
```

File: contracts/src/v0.8/ccip/test/capability/CCIPConfig.t.sol (hunk 10 above)
Comment (elatoskinas) [reply contracts/src/v0.8/ccip/test/capability/CCIPConfig.t.sol:48]: Ah that's unfortunate, could we add a TODO to replace it? Since we are planning an OZ v5 upgrade

File: contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol (hunk 5 above)
Comment (elatoskinas) [reply contracts/src/v0.8/ccip/capability/CCIPConfigArraysValidation.sol]: Current approach also makes sense to separate the empty error. Enum approach sounds good if there is a significant contract size reduction

File: contracts/src/v0.8/shared/util/SortedSetValidationUtil.sol [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    }
+  }
+
+  /// @notice Validates that a given set is sorted and has unique elements.
+  /// @dev Iterates through the array to check that each element is greater than the previous.
+  /// @param set The array of bytes32 to validate.
+  /// @custom:revert NotASortedSet If any element is not greater than its predecessor or if any two consecutive elements are equal.
+  function _checkIsValidSet(bytes32[] memory set) private pure {
+    for (uint256 i = 1; i < set.length; ++i) {
+      if (set[i] < set[i - 1] || set[i] == set[i - 1]) {
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/util/SortedSetValidationUtil.sol]: nit: gas optimization:

```suggestion
      if (set[i] <= set[i - 1]) {
```

File: contracts/src/v0.8/shared/util/SortedSetValidationUtil.sol [hunk 14]
```diff
@@ -0,0 +1,59 @@
+// SPDX-License-Identifier: BUSL-1.1
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/util/SortedSetValidationUtil.sol:1]: Since we are making this change against `contracts/shared`, we should:

1. Open a PR against [chainlink](https://github.com/smartcontractkit/chainlink) core repo, and place it in contracts/shared
2. Merge the PR in chainlink core
3. Cherry-pick the commit back to `ccip`

Example:
  * PR: https://github.com/smartcontractkit/chainlink/pull/13428
  * CCIP lib that was merged: [EnumerableMapAddresses](https://github.com/smartcontractkit/ccip/blob/ccip-develop/contracts/src/v0.8/shared/enumerable/EnumerableMapAddresses.sol)
  * `git cherry-pick` from `chainlink` origin for the PR that was merged in

File: contracts/src/v0.8/ccip/test/capability/CCIPConfig.t.sol (hunk 10 above)
Comment (0xsuryansh) [reply contracts/src/v0.8/ccip/test/capability/CCIPConfig.t.sol:48]: Done 👍🏼

File: contracts/src/v0.8/shared/util/SortedSetValidationUtil.sol (hunk 14 above)
Comment (0xsuryansh) [reply contracts/src/v0.8/shared/util/SortedSetValidationUtil.sol:1]: Understood, let me do this!
